OracleAnalytics/dockerfiles/6.4/createAndStartDomain.sh

```diff
@@ -0,0 +1,219 @@
+DOMAIN_HOME=$DOMAINS_DIR/${DOMAIN_NAME:-bi}
+
+# Use a touch file controlled by this script to ensure a container restart will know to start a successfully created domain.
+# But don't allow start for a failed configuration (we could delete DOMAIN_HOME and retry instead?)
+domainCheckFile=$DOMAIN_HOME/domainready.txt
+if [ ! -f "$domainCheckFile" ] && [ -d "$DOMAIN_HOME" ]; then
+  echo "An attempt was made to start a container with an incomplete DOMAIN_HOME"
+  exit 2
+fi
+
+# Load parameters from various files, overriding any environment variables.
+# Note we take care to not export these, so they are not visible to child processes
+loadParametersFile() {
+  file=$1/$2
+  if [ -f "$file" ]; then
+    if [ "txt" == "${file##*.}" ]; then
+      parameters=$(xargs < "$file")
+      log "Loaded from $file: $parameters"
+      eval "$(sed -e "s/=\(.*\)/='\1'/g" < "$file")"
+    else
+      parameterKey=$(echo "$2" | tr /a-z/ /A-Z/ | tr / _)
+      parameterValue=$(cat "$file")
+      parameter="$parameterKey='$parameterValue'"
+      log "Loaded from $file: $parameter"
+      eval "$parameter"
+    fi
+  fi  
+}
+
+parameterFileDir=/run/secrets
+parameterFiles=(config.txt admin.txt db.txt schema.txt admin/username admin/password db/username db/password schema/password)
+for parameterFile in "${parameterFiles[@]}"; do
+  loadParametersFile "$parameterFileDir" "$parameterFile"
+done
+
+# validate that all required parameters have been set somehow
+validateMandatoryParameter() {
+  key=$1
+  if [ -z "${!key}" ]; then
+    echo "$key"
+  fi
+}
+
+# TODO only ORACLE_HOME/DOMAIN_NAME/DOMAINS_DIR is mandatory once domain is configured
+mandatoryParameters=(ORACLE_HOME DOMAIN_NAME DOMAINS_DIR ADMIN_USERNAME ADMIN_PASSWORD DB_HOST DB_PORT DB_SERVICE DB_USERNAME DB_PASSWORD SCHEMA_PREFIX SCHEMA_PASSWORD)
+missingMandatoryParameters=()
+for mandatoryParameter in "${mandatoryParameters[@]}"; do
+  missingParam=$(validateMandatoryParameter "$mandatoryParameter")
+
+  if [[ ${missingParam} ]]; then
+    missingMandatoryParameters+=("${missingParam}")
+  fi
+done
+
+if (( ${#missingMandatoryParameters[@]} != 0 )); then
+  missingParametersJoin=$(IFS="," ; echo "${missingMandatoryParameters[*]}")
+  echo "The following mandatory parameters are not set: $missingParametersJoin"
+  exit 1
+fi
+
+for parameter in "${mandatoryParameters[@]}"; do
+  log "Parameter: $parameter=${!parameter}"
+done
+
+# TODO - in the case where docker secrets are used, it should be possible to remove the secret once the domain is created.
+#  However, that leaves no creds available for the dbping. Options: fail if no creds and DB_WAIT_TIMEOUT is set (i.e. not
+#  our responsibility to handle this on restart, b) work if schema password only is set (and then wait_for_db needs the prefix
+#  and not use 'as sysdba', c) as per b, but use the Connections API to recover a schema password from the domain (this makes
+#  more sense than (b).
```

Comment:
Either do this, or remove this from the file. Do not leave TODO items in published files.